### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by redmacdev1988**
Initial commit.

1) implemented circular array data structure to hold string urls. There are two implementations, one is for OOP, the other for FO. It will act as the carousel.

2) used the circular array in React. React display component Gallery simply checks to see if the structure has been sorted or re-fetched via useEffect.

### CODE REVIEW

File: src/CircularList/CircularlArrayFP.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+**/
+
+import ICircular from "./ICircular";
+
+let _data: readonly string [] = [];
+let _current: number = 0;
+let _counter: number = 0;
+
+const _updateWithItems = (newDataArr: string []) => {
+    _data = [...newDataArr];
```
Comment (neotan) [inline src/CircularList/CircularlArrayFP.ts]: generally, `_` is for private variables don't expect manually update in scope, e.g.:

```js
  const product  = {
      _id, // generated by database when inserting
      productName,
      _createdAt, // generated by database when inserting
    }

```

but declaring private variables is not a must, especially when doing so will hurt the readability.

**Review by neotan — CHANGES_REQUESTED** [review]

File: src/CircularList/CircularlArrayFP.ts (hunk 1 above)
Comment (redmacdev1988) [reply src/CircularList/CircularlArrayFP.ts]: removed _ for functionality that is returned for public usage.
removed _ for common data structure used.

Question - What do you mean by **private variables don't expect manually update in scope**. 
Like this?
```
let _data = [];
_data = [..._data, newValue]; // manually updated in scope.
```

File: src/CircularList/CircularlArrayFP.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+let numCounter: number = 0;
+
+const updateWithItems = (newDataArr: string []) => {
+    strArrData = [...newDataArr];
+    numCounter = newDataArr.length;
+}
+
+const insertAtFront = (newUrl: string): string => {
+    numCounter++;
+    strArrData = [newUrl, ...strArrData];
```
Comment (ginozhu) [inline src/CircularList/CircularlArrayFP.ts]: My two pence worth..
1. We should use deconstruction wisely as the performance is often bad
2. strArrData seems a bit odd, if it has to be a global stateful variable, it doesn't need to be reassigned everytime. If it doesnt, and you wanna keep FP's purity, we shall make it a parameter. (Also good for unit testing)
```
const insertAtFront = (strArrData: string [], newUrl: string)
```

**Comment by redmacdev1988** [comment]
https://user-images.githubusercontent.com/13611620/173977010-c70132fa-763b-4e5b-843b-7e4954e44527.mov

**Pull request merged**